harness/determined/ipc.py

```diff
@@ -439,3 +443,216 @@ def barrier(self, message: Any = None) -> Any:
 
     def close(self) -> None:
         self.socket.close()
+
+
+class PIDServer:
```

Comment:
I really admire the craftsmanship! But my take would be a http server with 
1. /hello request handler, submitting a pid
2. /shutdown request handler, informing about a [graceful] end
3. check_pids as a background thread
4. keepalive doesn't seem necessary

Client will make corresponding http requests using `requests`.

Reply:
What benefit would that strategy offer over the current strategy?

Reply:
no need to change it, this is great. it's just my opinion that http request/reply pattern would be more concise with a negligible loss of efficiency.

Reply:
Revisiting this after more thought and after discussion with @azhou-determined about the launch layer.

We realized it's really strange that horovod requires this sort of workaround.  I asked in the horovod slack; they confirmed it's not the intended behavior.  There's an upstream fix in the works (https://github.com/horovod/horovod/pull/3060).

It will of course be a while after the fix lands and before we can expect all customer images to contain fixed versions of horovod.  Therefore the PIDServer is going to stay for now, but I do think it will have a rather short lifetime.  Which is good, because it adds a lot of unnecessary complexity to the launch layer, which is meant to be user-customizable.